File: mobile/lib/services/socket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as IO;
import '../models/alarm.dart';

class SocketService {
  IO.Socket? _socket;
  final List<Function(Alarm)> _alarmListeners = [];

  void connect(String serverUrl) {
    disconnect();

    print('Connecting to socket: $serverUrl');

    _socket = IO.io(
      '$serverUrl/app',
      IO.OptionBuilder()
          .setTransports(['websocket'])
          .enableReconnection()
          .setReconnectionDelay(1000)
          .setReconnectionAttempts(10)
          .build(),
    );

    _socket?.on('connect', (_) {
      print('Socket connected');
      _socket?.emit('subscribe', {'app': 'mobile'});
    });

    _socket?.on('connected', (data) {
      print('Server acknowledged connection: $data');
    });

    _socket?.on('subscribed', (data) {
      print('Subscribed: $data');
    });

    _socket?.on('new_alarm', (data) {
      print('New alarm received: $data');
      try {
        final alarm = Alarm.fromJson(data);
        _notifyAlarmListeners(alarm);
      } catch (e) {
        print('Error parsing alarm: $e');
      }
    });

    _socket?.on('disconnect', (_) {
      print('Socket disconnected');
    });

    _socket?.on('connect_error', (error) {
      print('Connection error: $error');
    });

    _socket?.connect();
  }

  void disconnect() {
    _socket?.disconnect();
    _socket?.dispose();
    _socket = null;
  }

  void addAlarmListener(Function(Alarm) listener) {
    _alarmListeners.add(listener);
  }

  void removeAlarmListener(Function(Alarm) listener) {
    _alarmListeners.remove(listener);
  }

  void _notifyAlarmListeners(Alarm alarm) {
    for (var listener in _alarmListeners) {
      listener(alarm);
    }
  }

  bool get isConnected => _socket?.connected ?? false;
}

File: mobile/src/services/socketService.js
import io from 'socket.io-client';

let socket = null;
let alarmListeners = [];

export const initializeSocket = (serverUrl) => {
  if (socket) {
    socket.disconnect();
  }

  console.log('Connecting to server:', serverUrl);

  socket = io(serverUrl + '/app', {
    transports: ['websocket'],
    reconnection: true,
    reconnectionDelay: 1000,
    reconnectionAttempts: 10
  });

  socket.on('connect', () => {
    console.log('Connected to server');
    socket.emit('subscribe', { app: 'mobile' });
  });

  socket.on('connected', (data) => {
    console.log('Server acknowledged connection:', data);
  });

  socket.on('subscribed', (data) => {
    console.log('Subscribed:', data);
  });

  socket.on('new_alarm', (alarm) => {
    console.log('New alarm received:', alarm);
    notifyAlarmListeners(alarm);
  });

  socket.on('disconnect', () => {
    console.log('Disconnected from server');
  });

  socket.on('connect_error', (error) => {
    console.error('Connection error:', error);
  });

  return socket;
};

export const disconnectSocket = () => {
  if (socket) {
    socket.disconnect();
    socket = null;
  }
};

export const getSocket = () => socket;

export const addAlarmListener = (callback) => {
  alarmListeners.push(callback);
  return () => {
    alarmListeners = alarmListeners.filter(cb => cb !== callback);
  };
};

const notifyAlarmListeners = (alarm) => {
  alarmListeners.forEach(callback => callback(alarm));
};
